### ISSUE TITLE
Introduce new parameters to amber99sb_obc.xml

### ISSUE THREAD

**Issue by ghost** [openmm/openmm#2675]
Hello, 
My protein has about 4 NAG (N-Acetyl-beta-D-glucosamine) molecules. I need to optimize the protein with an implicit water model.

I edited the amber99sb.xml  to include the parameters of the NAG (angles, bonds, ..etc)  and optimization in vacuum worked fine. 

but when using the implicit water model (amber99sb_obc.xml) I get this error 

`ValueError: GBSAOBCForce: No parameters defined for atom type 1978`
I am new to openMM and have no idea from where could I get those parameters needed for NAG.

**Comment by F-AH** [comment]
more specifically I need to know how to get the scale parameter in the amber99sb_obc.xml file for the new atom types introduced to the ff file.

**Comment by peastman** [comment]
I'm not sure how Amber typically selects implicit solvent parameters for small molecules.  Where did you get the rest of the force field parameters from?

**Comment by jchodera** [comment]
@peastman: This description is relevant: https://github.com/openmm/openmmforcefields/issues/84#issuecomment-593110961

**Comment by F-AH** [comment]
> I'm not sure how Amber typically selects implicit solvent parameters for small molecules.  Where did you get the rest of the force field parameters from?

I got the rest of the parameters from this website 
[https://atb.uq.edu.au/molecule.py?molid=465793&fbclid=IwAR301Pt3fU-9Vtp9YOc8Ce72646ckXqU42bB-5Sh_T9jPX4j_7YM0Rt_NH4#panel-md](url)

**Comment by jchodera** [comment]
> I'm not sure how Amber typically selects implicit solvent parameters for small molecules.

The LEaP tool from AmberTools is responsible for [assigning GBSA parameters during `prmtop` writing](https://github.com/choderalab/ambermini/blob/8af922f71376aad6b01c876e4cb8e3fe3639e88c/leap/src/leap/unitio.c#L6061-L6293). This code was ported to [openmm](https://github.com/openmm/openmm/blob/master/wrappers/python/simtk/openmm/app/internal/customgbforces.py), but for some reason I'm unsure of, we only use it to _check_ parameters, not assign them.

It should be relatively straightforward just to extend `ForceField` to allow GBSA parameters to be assigned using this code via an optional argument. I'm not quite sure why we haven't done that before.

**Comment by peastman** [comment]
It could even be done in the XML file: include a script that calls that code to assign parameters.  I'll mark this as a feature request.

In the mean time, your easiest approach might be to use leap to build the system.  It will produce a prmtop file that you can load into OpenMM with AmberPrmtopFile.

**Comment by jchodera** [comment]
> It could even be done in the XML file: include a script that calls that code to assign parameters.

Great idea!

**Comment by F-AH** [comment]
> It could even be done in the XML file: include a script that calls that code to assign parameters.  I'll mark this as a feature request.
> In the mean time, your easiest approach might be to use leap to build the system.  It will produce a prmtop file that you can load into OpenMM with AmberPrmtopFile.

thanks a lot for your help

**Comment by F-AH** [comment]
> In the mean time, your easiest approach might be to use leap to build the system. It will produce a prmtop file that you can load into OpenMM with AmberPrmtopFile.

I tried to use leap to build the system and tried to optimize the protein structure using different solvation models (GBn, GBn2, HCT, OBC1, OBC2) 
(#note: I am running the simulation using CPUs )
for HCT, OBC1, OBC2
I got this error, 
`Exception: Particle coordinate is nan`
while for GBn and GBn2, 
I got this, 
`ValueError: Radii must be between 1 and 2 Angstroms for neck lookup`

I need to know what I am missing?  or what I am doing wrong?

also here is the script that I am using, 
`## Created by Leela S. Dodda for LigParGen Tutorials
## Date Aug 5, 2016
from simtk.openmm import app, KcalPerKJ
import simtk.openmm as mm
from simtk import unit as u
from sys import stdout, exit, argv
from simtk.openmm.app import *
from simtk.openmm import *
from simtk.unit import *
from sys import stdout


def Minimize(simulation,iters=0):
    print("start minimize")
    simulation.minimizeEnergy()
    print("done minimize")
    position = simulation.context.getState(getPositions=True).getPositions()
    energy = simulation.context.getState(getEnergy=True).getPotentialEnergy()
    app.PDBFile.writeFile(simulation.topology, position,
                          open('opt_'+name, 'w'))
    print('Energy at Minima is %3.3f kcal/mol' % (energy._value * KcalPerKJ))
    return simulation

print("load prmtop")   
prmtop = AmberPrmtopFile('prot.prmtop')
print("load inpcrd")   
inpcrd = AmberInpcrdFile('prot.inpcrd')
temperature = 298.15 * u.kelvin
print("create the system")
system = prmtop.createSystem(implicitSolvent=HCT, nonbondedMethod=NoCutoff, nonbondedCutoff=1*nanometer, constraints=None)
print("create the integrator object")
integrator = mm.LangevinIntegrator(temperature, 1 / u.picosecond,  0.0005 * u.picoseconds)
print("create the simulation object")
simulation = Simulation(prmtop.topology, system, integrator)
print("get the position from icpcrd")
simulation.context.setPositions(inpcrd.positions)
simulation = Minimize(simulation,10000)

`

**Comment by peastman** [comment]
See https://github.com/openmm/openmm/wiki/Frequently-Asked-Questions#nan which discusses that error message.  In this case it looks like either there must be something wrong with the way the forces are defined (possibly some inappropriate parameter values?), or else the initial conformation has a clash that's too severe for the minimizer to resolve.

Here are two useful tests to narrow it down.

First, identify if there are particular atoms causing the problems.  Query the forces, then search for ones with very large magnitude.  Look at those particles in a visualizer to see if there's an obvious problem with them.

```python
state = simulation.context.getState(getForces=True)
forces = state.getForces().value_in_unit(u.kilojoules_per_mole/u.nanometer)
for i,f in enumerate(forces):
  if u.norm(f) > 1e5:
    print(i, f)
```

Second, identify which force objects are causing the problems.  Before you create your Simulation put every force into a separate group:

```python
for i,f in enumerate(system.getForces()):
  f.setForceGroup(i)
```

Then you can query the energy of each one:

```python
for f in system.getForces():
  state = simulation.context.getState(getEnergy=True, groups={f.getForceGroup{}})
  print(f, state.getPotentialEnergy()
```

**Comment by F-AH** [comment]
> See https://github.com/openmm/openmm/wiki/Frequently-Asked-Questions#nan which discusses that error message.  In this case it looks like either there must be something wrong with the way the forces are defined (possibly some inappropriate parameter values?), or else the initial conformation has a clash that's too severe for the minimizer to resolve.
> Here are two useful tests to narrow it down.
> First, identify if there are particular atoms causing the problems.  Query the forces, then search for ones with very large magnitude.  Look at those particles in a visualizer to see if there's an obvious problem with them.
> state = simulation.context.getState(getForces=True)
> forces = state.getForces().value_in_unit(u.kilojoules_per_mole/u.nanometer)
> for i,f in enumerate(forces):
>   if u.norm(f) > 1e5:
>     print(i, f)
> Second, identify which force objects are causing the problems.  Before you create your Simulation put every force into a separate group:
> for i,f in enumerate(system.getForces()):
>   f.setForceGroup(i)
> Then you can query the energy of each one:
> for f in system.getForces():
>   state = simulation.context.getState(getEnergy=True, groups={f.getForceGroup{}})
>   print(f, state.getPotentialEnergy()

sorry for late reply, it was the clashes between cofactors in the protien structure which caused this error. 
using amber prepared prmtop and solvation HCT model worked perfectly with me 

thx a lot

**Comment by peastman** [comment]
Great, glad you figured it out!  Can this issue be closed?

By the way, HCT is one of the oldest and least accurate of the Amber solvation models.  One of the OBC or GBn models is probably a better choice.

**Comment by F-AH** [comment]
> Great, glad you figured it out!  Can this issue be closed?
> By the way, HCT is one of the oldest and least accurate of the Amber solvation models.  One of the OBC or GBn models is probably a better choice.

thx a lot for your help

**Comment by jchodera** [comment]
@peastman : Did we ever implement the feature described here to assign GBSA parameters via a Python script appended to the ffxml file?
https://github.com/openmm/openmm/issues/2675#issuecomment-628096198

**Comment by peastman** [comment]
Not yet.  I'll reopen this.  Thanks for point that out.

**Comment by peastman** [comment]
Fixed by #3214.
